File: day02/part1.py

result = 0
with open("day02/input.txt") as f:
    input = f.readline()
    ranges = input.split(",")
    for r in ranges:
        start, end = r.split("-")
        for i in range(int(start), int(end) + 1, ):
            str_i = str(i)
            if len(str_i) % 2 == 0 and str_i[0:len(str_i)//2] == str_i[len(str_i)//2:]:
                result += i
print(result)

File: day04/part1.pl
use v5.35;

open(my $in,  "<",  "day04/input.txt") or die "file not found";
my @rolls_layout = [];
my $height = 0;
my $width = 0;
while (<$in>) { 
    my @row_arr = split("", $_);
    $width = $#row_arr + 1;
    $rolls_layout[$height] = \@row_arr;
    $height++;
}

my @dirs = ([-1, -1], [-1, 0], [-1, 1], [0, -1], [0, 1], [1, -1], [1, 0], [1, 1]);
my $result = 0;
for (my $i = 0; $i < $height; $i++) {   
    for (my $j = 0; $j < $width; $j++) {
        my $neighbours = 0;
        foreach(@dirs) {
            my ($dx, $dy) = @{ $_ };
            my $nx = $j + $dx;
            my $ny = $i + $dy;

            if (0 <= $nx < $width && 0 <= $ny < $height && $rolls_layout[$ny][$nx] =~ '@') {
                $neighbours+=1;
            }
        }
        $result++ if $rolls_layout[$i][$j] =~ '@' && $neighbours < 4;
    }
}
print $result;
